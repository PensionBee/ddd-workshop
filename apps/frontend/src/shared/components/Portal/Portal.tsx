import { useLayoutEffect, useState } from "react";
import { createPortal } from "react-dom";

export type PortalProps = {
  children: React.ReactNode;
  wrapperId?: string;
};

export const Portal: React.FC<PortalProps> = ({
  children,
  wrapperId = "portal-root",
}) => {
  const [wrapperElement, setWrapperElement] = useState<HTMLElement | null>(
    null
  );

  useLayoutEffect(() => {
    const wrapperElement = document.getElementById(wrapperId);
    if (wrapperElement) {
      setWrapperElement(wrapperElement);
    } else {
      const createdWrapperElement = createWrapperAndAppendToBody(wrapperId);
      setWrapperElement(createdWrapperElement);

      return () => {
        createdWrapperElement.parentNode &&
          createdWrapperElement.parentNode.removeChild(createdWrapperElement);
      };
    }
  }, [wrapperId]);

  if (wrapperElement === null) return null;

  return createPortal(children, wrapperElement);
};

const createWrapperAndAppendToBody = (wrapperId: string) => {
  const wrapperElement = document.createElement("div");
  wrapperElement.setAttribute("id", wrapperId);
  document.body.appendChild(wrapperElement);
  return wrapperElement;
};
