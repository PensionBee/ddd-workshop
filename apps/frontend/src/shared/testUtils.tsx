import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
