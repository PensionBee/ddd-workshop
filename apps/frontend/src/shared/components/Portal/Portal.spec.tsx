import { render } from "@shared/testUtils";
import { Portal } from "./Portal";

describe("Portal", () => {
  const PORTAL_CONTENT = "Portal Content";

  describe("Default", () => {
    it("should render inside the default portal", () => {
      const { asFragment } = render(<Portal>{PORTAL_CONTENT}</Portal>);
      const portalRoot = document.querySelector("#portal-root");
      expect(portalRoot?.innerHTML).toBe(PORTAL_CONTENT);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("With wrapperId", () => {
    it("should render inside the specified wrapper", () => {
      const PORTAL_KEY = "portal-key";
      const { asFragment } = render(
        <Portal wrapperId={PORTAL_KEY}>{PORTAL_CONTENT}</Portal>
      );
      const portalRootFromKey = document.querySelector(`#${PORTAL_KEY}`);
      expect(portalRootFromKey?.innerHTML).toBe(PORTAL_CONTENT);
      expect(asFragment()).toMatchSnapshot();
    });

    it("should create wrapper if not exists", () => {
      const PORTAL_KEY = "portal-key";
      const { asFragment } = render(
        <Portal wrapperId={PORTAL_KEY}>{PORTAL_CONTENT}</Portal>
      );
      const portalRootFromKey = document.querySelector(`#${PORTAL_KEY}`);
      expect(portalRootFromKey).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });

    it("should use wrapper if already exists", () => {
      const PORTAL_KEY = "portal-key";
      const { asFragment } = render(
        <>
          <Portal wrapperId={PORTAL_KEY}>{PORTAL_CONTENT}</Portal>
          <div id={PORTAL_KEY} />
        </>
      );
      const portalRootFromKey = document.querySelector(`#${PORTAL_KEY}`);
      expect(portalRootFromKey).toBeInTheDocument();
      expect(portalRootFromKey?.innerHTML).toBe(PORTAL_CONTENT);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
