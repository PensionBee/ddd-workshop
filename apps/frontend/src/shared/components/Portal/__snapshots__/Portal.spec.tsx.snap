// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Portal > Default > should render inside the default portal 1`] = `<DocumentFragment />`;

exports[`Portal > With wrapperId > should create wrapper if not exists 1`] = `<DocumentFragment />`;

exports[`Portal > With wrapperId > should render inside the specified wrapper 1`] = `<DocumentFragment />`;

exports[`Portal > With wrapperId > should use wrapper if already exists 1`] = `
<DocumentFragment>
  <div
    id="portal-key"
  >
    Portal Content
  </div>
</DocumentFragment>
`;

exports[`Portal Default should render inside the default portal 1`] = `<DocumentFragment />`;

exports[`Portal With wrapperId should create wrapper if not exists 1`] = `<DocumentFragment />`;

exports[`Portal With wrapperId should render inside the specified wrapper 1`] = `<DocumentFragment />`;

exports[`Portal With wrapperId should use wrapper if already exists 1`] = `
<DocumentFragment>
  <div
    id="portal-key"
  >
    Portal Content
  </div>
</DocumentFragment>
`;
